### DISCUSSION TITLE
How to Implement Nonlinear Stiffness Based on Deformation?

### DISCUSSION THREAD

**Discussion by cuicw** [GazzolaLab/PyElastica#344] [Q&A] [answered]
I'm currently studying with `pyelastica` and am trying to implement a feature where I need to setup a nonlinear stiffness that changes with deformation.

**Comment by armantekinalp** [comment]
Do you have the stiffness model? Which stiffness are you talking about bending, axial stretch, shear etc?

**Reply by cuicw** [reply]
Yes, I have a nonlinear bending stiffness model with curvature as the independent variable. I'm currently not sure in which module to develop it.

**Reply by armantekinalp** [reply]
Seems like you are only changing bending stiffness. I want to confirm. In your relation is it still `internal_torques = f(\kappa) (\kappa-\rest_kappa)` where `f(\kappa)` is your nonlinear bending stiffness?

**Reply by cuicw** [reply]
Yes

**Reply by armantekinalp** [reply, marked as answer]
There are two options:

1. Easier option. You can write a forcing function and modify the `bend_matrix` at each time step. Note that both `bend_matrix` and `kappa` are members of the rod. Now here there is a catch with this approach, since in time-step algorithm all external_force functions computed after the internal_forces/internal_torques, updated `bend_matrix` will be used in the next time-step. If you expect significant changes between `kappa` values in between time-steps then I suggest you try the second approach.

2. Relatively harder option. You can modify the functions where internal torques are computed, but this requires changing some back end code. I suggest clone PyElastica in your local system and change the following functions. [Before this line](https://github.com/GazzolaLab/PyElastica/blob/master/elastica/rod/cosserat_rod.py#L888) you can modify the `bend matrix` accordingly. Note that this is a numba function so it would not support some Numpy functions, make sure to check this [link](http://numba.pydata.org/numba-doc/dev/reference/numpysupported.html).

**Reply by cuicw** [reply]
Thank you.
